### PR TITLE
[BFT] Error handling for engines

### PR DESCRIPTION
**Pull request by durkmurder**
### Context

This PR unifies error handling and they way we process message delivered from network layer or other components.
In nutshell:
- `Process/Submit` never returns `IncompatibleInputTypeError` since it expected byzantine inputs
- `ProcessLocal/SubmitLocal` doesn't tolerate `IncompatibleInputTypeError` and will propagate error to caller since input is trusted and it's unexpected. 

This PR has lots of boilerplate logic, I am trying to condense it into reusable component something like `engineBase` or `engineManager`. On other hand this can wait till we refactor `network.Engine` in favor of `network.MessageProcessor `

### CODE REVIEW

File: engine/consensus/matching/engine.go [hunk 1]
```diff
@@ -151,9 +160,10 @@ func (e *Engine) process(originID flow.Identifier, event interface{}) error {
 // HandleReceipt ingests receipts from the Requester module.
 func (e *Engine) HandleReceipt(originID flow.Identifier, receipt flow.Entity) {
 	e.log.Debug().Msg("received receipt from requester engine")
-	e.metrics.MessageReceived(metrics.EngineSealing, metrics.MessageExecutionReceipt)
-	e.pendingReceipts.Push(receipt)
-	e.inboundEventsNotifier.Notify()
+	err := e.process(originID, receipt)
```
Comment (AlexHentschel) [inline engine/consensus/matching/engine.go:163]: thanks for fixing this too

File: engine/consensus/matching/engine.go [hunk 2]
```diff
@@ -133,14 +133,23 @@ func (e *Engine) ProcessLocal(event interface{}) error {
...
+		if engine.IsIncompatibleInputTypeError(err) {
+			e.log.Warn().Msgf("%v delivered unsupported message %T through %v", originID, event, channel)
+			return nil
+		}
+		return fmt.Errorf("unexpected error while processing engine message: %w", err)
+	}
+	return nil
 }
 
 // process processes events for the matching engine on the consensus node.
```
Comment (AlexHentschel) [inline engine/consensus/matching/engine.go:147]:
```suggestion
// process events for the matching engine on the consensus node.
```

**Review by AlexHentschel — APPROVED** [review]
thanks for fixing this. Really great that you also improved the collection nodes. 🏅  

🚢

**Review by jordanschalm — APPROVED** [review]
Great work, thank you 🎸

**Comment by durkmurder** [comment]
bors merge

**Comment by durkmurder** [comment]
bors retry

**Pull request merged**